### PR TITLE
tests(adds tests, 100% code coverage for core)

### PR DESCRIPTION
**Pull request by joshribakoff**
- Adds tests (progress against #56).
- Please review & merge #48 first, then change this PR to use `master` as the base before reviewing it (this depends on changes introduced is the `visualizer` branch, namely the addition of the "store events" and the ability to observe the inner workings of the store, which I verify with these tests).

### CODE REVIEW

File: libs/core/src/lib/effect.spec.ts [hunk 1]
```diff
@@ -1,3 +1,121 @@
...
+import { setup } from './test-helper';
+
+it('spawns and tears down an empty effect', () => {
+  const effect: Effect<{}> = () => of();
+  setup({ value: {}, effect }, ({ expectObservable, effect$, storeEvent$ }) => {
+    expectObservable(effect$).toBe('|');
+    expectObservable(storeEvent$).toBe('(ab)', {
+      a: { event: 'spawn', name: 'root', type: 'effect' },
+      b: { event: 'teardown', name: 'root', type: 'effect' },
+    });
```
Comment (joshribakoff) [inline libs/core/src/lib/effect.spec.ts:12]: The idea w/ this style of tests is to verify what would otherwise be an "implementation detail", but is actually a contract I want to enforce, since I intend this `storeEvent$` stream to be part of our "public API", breaking changes should require a major version bump, and these tests should catch any issues like running effects more or less often than expected, or in different orders, even if the effect otherwise produces the same values.

I don't use the enums but rather add raw strings, because again its public API, and changes to the values in the enum should be considered breaking, so its undesirable to use an enum in the test where the test could easily be updated erroneously by using a refactoring tool. I'd like this test to fail, even if someone applies a "refactor", because its not really a refactor, its part of the public API & therefore it'd just be a behavior change not a refactor if it were to change.

File: libs/core/src/lib/sources.spec.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+it('sources data from subject', () => {
+  const value = { count$: new Subject() };
+  const effect: Effect<typeof value> = ({ sources }) => {
+    return sources.count$();
+  };
+  setup(
+    { value, effect },
+    ({ expectObservable, effect$, hot, storeEvent$ }) => {
+      hot('a').subscribe(value.count$);
```
Comment (joshribakoff) [inline libs/core/src/lib/sources.spec.ts:13]: @kylecannon any idea why changing this to cold breaks the test? It should be no different, in both cases we subscribe synchronously before the rest of the test runs.

File: libs/core/src/lib/sources.spec.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+          from: { type: 'subject', name: 'count$' },
+          to: { type: 'effect', name: 'root' },
+        },
+        c: {
+          type: 'value',
+          from: { type: 'subject', name: 'count$' },
+          to: { type: 'effect', name: 'root' },
+          value: 'a',
+        },
+        d: {
```
Comment (joshribakoff) [inline libs/core/src/lib/sources.spec.ts:29]: I suppose there are more edge cases even though we're at 100% coverage, in a follow-up PR I can add variations of this test where the child does not allow the events to bubble to the parent effect, etc. to make sure we permute more of the edge cases.

File: libs/core/src/lib/sinks.spec.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+it('sources data from subject', () => {
+  const value = { count$: new Subject() };
+  const effect: Effect<typeof value> = ({ sinks }) => {
+    return of('hello world').pipe(sinks.count$());
+  };
+  setup({ value, effect }, ({ expectObservable, effect$, storeEvent$ }) => {
+    expectObservable(effect$).toBe('(a|)', { a: 'hello world' });
+
+    expectObservable(storeEvent$).toBe('(abcde)', {
```
Comment (joshribakoff) [inline libs/core/src/lib/sinks.spec.ts:13]: Some overlap here in these tests. In the future we could maybe make these all unit tests, and extract the "integration tests" to live somewhere else & DRY it up. For now, it seemed better to write tests specifically for testing the sources & sinks, but to also test via the top level effects API / test helper for more verification & higher fidelity.
